File: ScaleformUI_Lua/src/scaleforms/BigFeed/BigFeedHandler.lua
BigFeedInstance = setmetatable({
    _sc = nil --[[@type Scaleform]],
    _title = "",
    _subtitle = "",
    _bodyText = "",
    _txn = "",
    _txd = "",
    _enabled = false,
    _rightAligned = false,
    _disabledNotifications = false,
}, BigFeedInstance)
BigFeedInstance.__index = BigFeedInstance
BigFeedInstance.__call = function()
    return "BigFeedInstance"
end

---@class BigFeedInstance
---@field public _sc Scaleform
---@field private _title string
---@field private _subtitle string
---@field private _bodyText string
---@field private _txn string
---@field private _txd string
---@field public _enabled boolean
---@field private _rightAligned boolean
---@field private _disabledNotifications boolean
---@field public Title fun(self:BigFeedInstance, title:string):string
---@field public Subtitle fun(self:BigFeedInstance, subtitle:string):string
---@field public BodyText fun(self:BigFeedInstance, body:string):string
---@field public Texture fun(self:BigFeedInstance, textureName:string, textureDictionary:string):table<string, string>
---@field public Enabled fun(self:BigFeedInstance, enabled:boolean):boolean
---@field public UpdateInfo fun(self:BigFeedInstance):nil
---@field public Update fun(self:BigFeedInstance):nil
---@field public Load fun(self:BigFeedInstance):nil
---@field public Dispose fun(self:BigFeedInstance):nil


---Sets the title of the BigFeedInstance, if no title is provided it will return the current title
---@param title? string
---@return string
function BigFeedInstance:Title(title)
    if title == nil then return self._title end
    self._title = tostring(title)
    if self._enabled then
        self:UpdateInfo()
    end

    return self._title
end

---Sets the subtitle of the BigFeedInstance, if no subtitle is provided it will return the current subtitle
---@param subtitle? string
---@return string
function BigFeedInstance:Subtitle(subtitle)
    if subtitle == nil then return self._subtitle end
    self._subtitle = tostring(subtitle)
    if self._enabled then
        self:UpdateInfo()
    end
    return self._subtitle
end

---Sets the body text of the BigFeedInstance, if no body text is provided it will return the current body text
---@param body? string
---@return string
function BigFeedInstance:BodyText(body)
    if body == nil then return self._bodyText end
    self._bodyText = tostring(body)
    if self._enabled then
        self:UpdateInfo()
    end
    return self._bodyText
end

---Sets the BigFeedInstance to be right aligned, if no state is provided it will return the current state
---@param rightAligned? boolean
---@return boolean
function BigFeedInstance:RightAligned(rightAligned)
    if rightAligned == nil then return self._rightAligned end
    self._rightAligned = ToBool(rightAligned)
    if self._enabled then
        self._sc:CallFunction("SETUP_BIGFEED", self._rightAligned)
        self:UpdateInfo()
    end
    return self._rightAligned
end

---Sets the texture of the BigFeedInstance, if no texture is provided it will return the current texture
---@param textureName? string
---@param textureDictionary? string
---@return table<string, string>
function BigFeedInstance:Texture(textureName, textureDictionary)
    if textureName == nil and textureDictionary == nil then return { txd = self._txd, txn = self._txn } end
    self._txn = textureName
    self._txd = textureDictionary
    if self._enabled then
        self._sc:CallFunction("SET_BIGFEED_IMAGE", textureDictionary, textureName)
        self:UpdateInfo()
    end
    return { txd = self._txd, txn = self._txn }
end

---Toggles the BigFeedInstance on or off, if a state is not provided it will return the current enabled state
---@param enabled? boolean
---@return boolean
function BigFeedInstance:Enabled(enabled)
    if enabled == nil then return self._enabled end
    self._enabled = ToBool(enabled)
    if enabled == true then
        self:Load()
        self._sc:CallFunction("SETUP_BIGFEED", self._rightAligned)
        self._sc:CallFunction("HIDE_ONLINE_LOGO")
        self._sc:CallFunction("FADE_IN_BIGFEED")
        if self._disabledNotifications then
            ThefeedCommentTeleportPoolOn()
        end
        self:UpdateInfo()
    else
        self._sc:CallFunction("END_BIGFEED")
        if self._disabledNotifications then
            ThefeedCommentTeleportPoolOff()
        end
        self:Dispose()
    end
    return self._enabled
end

function BigFeedInstance:DisableNotifications(bool)
    if bool == nil then
        return self._disabledNotifications
    else
        self._disabledNotifications = bool
    end
end

---Updates the information displayed on the BigFeedInstance
---@return nil
function BigFeedInstance:UpdateInfo()
    if self._sc == nil or not self._enabled then return end
    AddTextEntry("scaleform_ui_bigFeed", self._bodyText)
    BeginScaleformMovieMethod(self._sc.handle, "SET_BIGFEED_INFO")
    ScaleformMovieMethodAddParamTextureNameString("")
    BeginTextCommandScaleformString("scaleform_ui_bigFeed")
    EndTextCommandScaleformString_2()
    ScaleformMovieMethodAddParamInt(0)
    ScaleformMovieMethodAddParamTextureNameString(self._txn)
    ScaleformMovieMethodAddParamTextureNameString(self._txd)
    ScaleformMovieMethodAddParamTextureNameString(self._subtitle)
    ScaleformMovieMethodAddParamTextureNameString("")
    ScaleformMovieMethodAddParamTextureNameString(self._title)
    ScaleformMovieMethodAddParamInt(0)
    EndScaleformMovieMethod()
end

---Draws the BigFeedInstance on the screen
---@return nil
function BigFeedInstance:Update()
    if self._sc == nil or self._sc == 0 or not self._enabled then return end
    self._sc:Render2D()
end

---Loads the BigFeedInstance scaleform
---@return nil
function BigFeedInstance:Load()
    if self._sc ~= nil then return end
    self._sc = Scaleform.Request("GTAV_ONLINE")
    local timeout = 1000
    local start = GlobalGameTimer
    while not self._sc:IsLoaded() and GlobalGameTimer - start < timeout do Citizen.Wait(0) end
    self._sc:CallFunction("HIDE_ONLINE_LOGO")
end

---Disposes the BigFeedInstance scaleform
function BigFeedInstance:Dispose()
    self._sc:Dispose()
    self._sc = nil
end

File: ScaleformUI_Csharp/Scaleforms/BigFeed/BigFeedHandler.cs
using CitizenFX.Core;
using static CitizenFX.FiveM.Native.Natives;

namespace ScaleformUI.Scaleforms
{
    public class BigFeedHandler
    {
        internal ScaleformWideScreen _sc;
        private bool enabled;
        private string title = "";
        private string subtitle = "";
        private string bodyText = "";
        private string textureDictionary = "";
        private string textureName = "";

        /// <summary>
        /// Decides if the feed is centered or right aligned, *MUST* be called before Enabled
        /// </summary>
        public bool RightAligned { get; set; }
        /// <summary>
        /// If true disables all kind of notifications while the feed is showing
        /// </summary>
        public bool DisabledNotifications { get; set; }
        /// <summary>
        /// The title of the loading screen.
        /// </summary>
        public string Title
        {
            get => title;
            set
            {
                title = value;
                if (Enabled)
                {
                    UpdateInfo();
                }
            }
        }

        /// <summary>
        /// The subtitle of the loading screen.
        /// </summary>
        public string Subtitle
        {
            get => subtitle;
            set
            {
                subtitle = value;
                if (Enabled)
                {
                    UpdateInfo();
                }
            }
        }

        /// <summary>
        /// The description of the loading screen.
        /// </summary>
        public string BodyText
        {
            get => bodyText;
            set
            {
                bodyText = value;
                if (Enabled)
                {
                    UpdateInfo();
                }
            }
        }

        /// <summary>
        /// The Texture Dictionary (TXD) where the texture is loaded.
        /// </summary>
        public string TextureDictionary { get => textureDictionary; private set => textureDictionary = value; }

        /// <summary>
        /// The texture in the dictionary.
        /// </summary>
        public string TextureName { get => textureName; private set => textureName = value; }

        public bool Enabled
        {
            get => enabled;
            set
            {
                enabled = value;
                if (value)
                {
                    _sc.CallFunction("SETUP_BIGFEED", RightAligned);
                    _sc.CallFunction("HIDE_ONLINE_LOGO");
                    _sc.CallFunction("FADE_IN_BIGFEED");
                    if (DisabledNotifications)
                        ThefeedCommentTeleportPoolOn();
                    UpdateInfo();
                }
                else
                {
                    _sc.CallFunction("END_BIGFEED");
                    //_sc.CallFunction("END_BIGFEED");
                    if (DisabledNotifications)
                        ThefeedCommentTeleportPoolOff();
                }
            }
        }

        public BigFeedHandler() { Load(); }

        private async void Load()
        {
            if (_sc != null) return;
            _sc = new ScaleformWideScreen("GTAV_ONLINE");
            int timeout = 1000;
            int start = Main.GameTime;
            while (!_sc.IsLoaded && Main.GameTime - start < timeout) await BaseScript.Delay(0);
            _sc.CallFunction("HIDE_ONLINE_LOGO");
        }
        private void Dispose()
        {
            _sc.Dispose();
            _sc = null;
        }

        public void UpdatePicture(string txd, string txn)
        {
            TextureDictionary = txd;
            TextureName = txn;
            if (!enabled) return;
            _sc.CallFunction("SET_BIGFEED_IMAGE", txd, txn);
        }

        private void UpdateInfo()
        {
            if (!enabled) return;
            AddTextEntry("scaleform_ui_bigFeed", bodyText);
            BeginScaleformMovieMethod(_sc.Handle, "SET_BIGFEED_INFO");
            PushScaleformMovieFunctionParameterString(string.Empty);
            BeginTextCommandScaleformString("scaleform_ui_bigFeed");
            EndTextCommandScaleformString_2();
            PushScaleformMovieFunctionParameterInt(0);
            PushScaleformMovieFunctionParameterString(textureDictionary);
            PushScaleformMovieFunctionParameterString(textureName);
            PushScaleformMovieFunctionParameterString(subtitle);
            PushScaleformMovieFunctionParameterString(string.Empty);
            PushScaleformMovieFunctionParameterString(title);
            PushScaleformMovieFunctionParameterInt(0);
            EndScaleformMovieMethod();
        }

        internal async void Update()
        {
            if (enabled)
                _sc.Render2D();
        }
    }
}
